### PR TITLE
KAFKA-12984: make AbstractStickyAssignor resilient to invalid input, utilize generation in cooperative, and fix assignment bug

### PR DESCRIPTION
**Pull request by ableegoldman**
The primary goal of this PR is to address the problem we've seen in the wild in which the ConsumerCoordinator fails to update its SubscriptionState and ultimately feeds invalid `ownedPartitions` data as input to the assignor. Previously the assignor would detect that something was wrong and just throw an exception, now we make several efforts to detect this earlier in the assignment process and then fix it if possible, and work around it if not.

Specifically, this PR does a few things:
1) Bring the `generation` field back to the CooperativeStickyAssignor so we don't need to rely so heavily on the ConsumerCoordinator properly updating its SubscriptionState after eg falling out of the group. The plain StickyAssignor always used the generation since it had to, so we just make sure the CooperativeStickyAssignor has this tool as well
2) In case of unforeseen problems or further bugs that slip past the `generation` field safety net, the assignor will now explicitly look out for partitions that are being claimed by multiple consumers as owned in the same generation. Such a case should never occur, but if it does, we have to invalidate this partition from the `ownedPartitions` of both consumers, since we can't tell who, if anyone, has the valid claim to this partition.
3) Fix a subtle bug that I discovered while writing tests for the above two fixes: in the constrained algorithm, we compute the exact number of partitions each consumer should end up with, and keep track of the "unfilled" members who must -- or _might_ -- require more partitions to hit their quota. The problem was that members at the `minQuota` were being considered as "unfilled" even after we had already hit the maximum number of consumers allowed to go up to the `maxQuota`, meaning those `minQuota` members could/should not accept any more partitions beyond that. I believe this was introduced in [#10509](https://github.com/apache/kafka/pull/10509), so it shouldn't be in any released versions and does not need to be backported.

### CODE REVIEW

**Comment by ableegoldman** [comment]
ready for review @dajac @guozhangwang @hachikuji  @showuon

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: This part I just moved here to keep things up to date as we go, before we were clearing them after the loop

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: This is fix #2 -- if we somehow still get multiple consumers claiming a partition in the same generation, we have to consider both invalid and remove it from their `ownedPartitions`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Strictly speaking this should never ever happen, even if we do get these "impossible" doubly-claimed partitions, we're also removing them from the `ownedPartitions` above (that's fix #2). But I put in a safeguard just in case, it shouldn't hurt (performance-wise we should generally not even enter this loop since `partitionsWithMultiplePreviousOwners` should almost always be empty

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: This is part of fix #3 -- basically we have to handle these members separately, once they get up to `minQuota` they are only "potentially" unfilled. Once the last member allowed reaches `maxQuota`, all of these `minQuota` members are suddenly considered filled. cc @showuon

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (ableegoldman) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:560]: Saw this fail sometimes when running locally. This PR definitely should not have significantly affected the performance of this test in any way, since the only substantial changes are in the constrained assignment algorithm, while this test targets the general case of unequal subscriptions. It's not too surprising to me that this would be running up against the time limit anyways, since we've known that the general assignment algorithm is complicated and doesn't scale too well.So I felt it was reasonable to give it a bit more time. 

Alternatively we could just scale back a bit with the number of partitions and/or consumers. WDYT @showuon?

Comment (showuon) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:560]: Increase the timeout makes sense to me.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Because we cleared them earlier now, the `membersWithOldGeneration` is not necessary any more. We can just iterate `membersOfCurrentHighestGeneration` here.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: add a space after "multiple". i.e. `despite being claimed by multiple[ ]`

**Review by showuon — COMMENTED** [review]
@ableegoldman , Thanks for the fix (of my bug :) )! have a quick look and left some comments. Will look again tomorrow. Thanks.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Should we log the generation number `maxGeneration` ?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: `expectedAssignedCount` is not used any more.

File: clients/src/main/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignor.java
Comment (jacky1193610322) [inline clients/src/main/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignor.java]: Should we create a schema class, not direct ByteBuffer for adding a field easily in the future?

Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/CooperativeStickyAssignor.java]: Yeahh...will do

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Mm good catch

**Comment by ableegoldman** [comment]
Thanks for the reviews so far, all comments should be addressed. Migrated the encoded `generation` userData over to an actual schema in case we ever want or need to add more fields/metadata for the assignor to use.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:301]: I'm thinking we can have a test for package scope `partitionsTransferringOwnership` in `AbstractStickyAssignorTest`.  I found we didn't test it before. We can verify the doubly assigned partitions and other revoked partitions are put into `partitionsTransferringOwnership` correctly.

Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:301]: That's a good idea. But WDYT about doing this in a small followup PR? Partly to keep this PR from growing too long (it's already much longer than I originally intended), and partly so we can get this one merged ASAP since it's a blocker.

Actually, would you be interested in working on that test (or tests) yourself, after this PR can be merged? At this point I think you're probably the most familiar with this code outside of myself

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:301]: `KAFKA-13046: Improve the test coverage for stickyAssignor` is created.
Let me handle it! :)

**Review by showuon — APPROVED** [review]
Have a 2nd review. LGTM! Thanks for the fix to make the cooperative-sticky assignor more reliable!

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:131]: If the current member's generation is available but `<` maxGeneration, should we also clear it from the `consumerToOwnedPartitions` map? I think the passed in `subscriptions` is not sorted by the generations right?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Honestly it took me quite a while to understand the fix :P After understanding that I think maybe it's better to rename these two collections more explicitly:

1) `unfilledMembers` -> `MembersWithLessThanMinQuotaPartitions`.
2) `potentiallyUnfilledMembersAtMinQuota` -> `MembersWithExactMinQuotaPartitions`.

And also (since the maxQuota is always either == minQuota or minQuota + 1):

3) `expectedNumMembersAssignedOverMinQuota` -> `expectedNumMembersWithMaxQuota` 
4) `numMembersAssignedOverMinQuota` -> `numMembersWithMaxQuota`

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: I'd suggest we remove this `potentiallyUnfilledMembersAtMinQuota.clear();` logic and just keep two expected numbers:

1) expectedNumMembersWithMaxQuota = totalPartitionsCount % numberOfConsumers;
2) expectedNumMembersWithMinQuota = numberOfConsumers - expectedNumMembersWithMaxQuota;

And then we can also remove the check in line 309, and at the end when we exhausted `unassignedPartitions`, just check

```
numMembersWithMaxQuota == expectedNumMembersWithMaxQuota &&
numMembersWithMinQuota == expectedNumMembersWithMinQuota
```

WDYT?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: nit: do you think we should log at ERROR since this is not expected really? Right now we would sort of "hide" such bugs and still be able to proceed silently; I feel we should shouting out such scenarios a bit louder in logs.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Same here: in our current code this should never happen, so what about log as ERROR?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Related to the one above: maybe we just check that 

```
numMembersWithMaxQuota == expectedNumMembersWithMaxQuota &&
numMembersWithMinQuota == expectedNumMembersWithMinQuota
```

And if not, log the full assignment as an ERROR?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java
Comment (guozhangwang) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignorTest.java:560]: +1

**Review by guozhangwang — COMMENTED** [review]
Thanks for the patch @ableegoldman ! I made a pass on it.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:131]: In that case, it's never added to `consumerToOwnedPartitions` in the first place. This map is not pre-filled, it gets populated inside this loop. So if its `< maxGeneration`, then we just insert an empty list into the map for that member's owned partitions

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Good point, yes I would absolutely want/hope a user would report this as a bug. Changed to ERROR

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Ack on the first two renamings, though I'd still want to prefix them with `unfilled` to emphasize that these structures only hold members that may potentially be assigned one or more partitions. ie, if `minQuota == maxQuota`, then `potentiallyUnfilledMembersAtMinQuota` should actually be empty, in which case `MembersWithExactMinQuotaPartitions` doesn't quite make sense. I'll clarify this in the comments as well.

Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: For the 3 & 4th suggested renamings, it's a bit subtle but this would actually be incorrect. In the case `minQuota == maxQuota`, the `expectedNumMembersAssignedOverMinQuota` variable will evaluate to 0, which would not make sense if it was called `expectedNumMembersWithMaxQuota`. Of course we could go through a tweak the logic for this case, but I'd prefer not to mix that into this PR. For now I'll just clarify in the comments for these variables.
(I did still rename them slightly to hopefully be more clear, and also in line with the new names of the other two data structures we renamed)

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: While I'm not really a fan of the `potentiallyUnfilledMembersAtMinQuota` logic (it's definitely awkward but I felt it was still the lesser evil in terms of complicating the code), I don't think we can get rid of it that easily. The problem is that when `minQuota != maxQuota`, and so far `currentNumMembersWithOverMinQuotaPartitions` < `expectedNumMembersWithOverMinQuotaPartitions`, then consumers that are filled up to exactly `minQuota` have to be considered potentially not yet at capacity since some will need one more partition, though not all. So this data structure is not just used to verify that everything is properly assigned after we've exhausted the `unassignedPartitions`, it's used to track which consumers can still receive another partition (ie, are "unfilled"). Does that make sense?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: I responded to the above comment as well, but specifically here I think that to just check on that condition requires us to make assumptions about the algorithm's correctness up to this point (and the correctness of its assumptions). But if those are all correct then we would never reach this to begin with, so it's better to directly look for any remaining `unassignedPartitions` -- it's a sanity check. 

But ack on bumping to ERROR

**Comment by ableegoldman** [comment]
Responded to your comments @guozhangwang , let me know if that all makes sense or if you have any more concerns that need to be addressed in this PR

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Ack.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Yes that makes sense, still this logic

```
if (numMembersAssignedOverMinQuota == expectedNumMembersAssignedOverMinQuota) {
                    potentiallyUnfilledMembersAtMinQuota.clear();
}
```

Seems only needed because we have the check in 309 (?) Say if we do not check that, but instead just check the expected numbers of consumers with minQuota and maxQuota is satisfied, then do we still need this?

**Comment by guozhangwang** [comment]
Just one more reply on the clearing logic, otherwise LGTM.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: > this logic Seems only needed because we have the check in 309 (?)

No, I don't think so. It should be for line 279:
```java
// to handle the case that when there are still unassignedPartition left, but no more members to be assigned.
if (unfilledMembersWithUnderMinQuotaPartitions.isEmpty() && unfilledMembersWithExactlyMinQuotaPartitions.isEmpty()) {
   throw new IllegalStateException("No more unfilled consumers to be assigned.");
```

In line 309, it is just an early error detect and log for it. Not related to `potentiallyUnfilledMembersAtMinQuota` (or now `unfilledMembersWithExactlyMinQuotaPartitions` members)

Comment (showuon) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: I tried to summarize both methods:

What @guozhangwang 's meaning is, we can "lazily" detect the issue after assigning all unassignedPartitions. we don't need to clear the `potentiallyUnfilledMembersAtMinQuota` here, because as the "original" variable naming said: they are "potentially unfilled members", just keep them there. We "should not" assign partitions to them in this case because we've reached `expectedNumMembersAssignedOverMinQuota`. 

But if somehow, after assigning unassignedPartitions to all unfilledMembers, there are still some unassignedPartitions left. We can just assign them to `potentiallyUnfilledMembersAtMinQuota`. And after running out the `unassignedPartitions`, we can check:
```
numMembersWithMaxQuota == expectedNumMembersWithMaxQuota &&
numMembersWithMinQuota == expectedNumMembersWithMinQuota
```

to do error handling. 

VS.

In @ableegoldman 's version , we find issue immediately and handle it. we computed the `potentiallyUnfilledMembersAtMinQuota` correctly (that's why we need to clear it). So, if the issue happened:  
> if somehow, after assigning unassignedPartitions to all unfilledMembers, there are still some unassignedPartitions left

We can try to get member from `potentiallyUnfilledMembersAtMinQuota` and then assign unassignedPartition to the member. If we can't get member from it (i.e. `potentiallyUnfilledMembersAtMinQuota` is empty), we throw exception directly.


Both ways can find errors when happened. Personally, I like Sophie's version more since it's much clear.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (showuon) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java:269]: @ableegoldman , I reviewed it again, and found we forgot to sort the `unfilledMembersWithExactlyMinQuotaPartitions` list here, to have deterministic result.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java
Comment (ableegoldman) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: Thanks @showuon and @guozhangwang , I think that all makes sense. One of my primary motivations was to keep all data structures at all times consistent with what they represent so they could always be relied upon to be used at any point. For that reason I also prefer to keep things as is, and clear the `potentiallyUnfilledMembersAtMinQuota` (now renamed to `nfilledMembersWithExactlyMinQuotaPartitions`) as soon as we have filled the last member who may be above `minQuota`, at which point all of the members at exactly `minQuota` are no longer considered "unfilled"

Comment (guozhangwang) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractStickyAssignor.java]: SGTM.

**Comment by guozhangwang** [comment]
LGTM!

**Review by showuon — APPROVED** [review]
LGTM! Thanks for the update!

**Comment by ableegoldman** [comment]
Some unrelated flaky test failures: `RaftEventSimulationTest.canRecoverAfterAllNodesKilled` and `ConsumerBounceTest.testCloseDuringRebalance()` (known to be flaky, has been failing often lately)

**Comment by ableegoldman** [comment]
Merged to trunk and cherrypicked back to 3.0 (cc @kkonstantine)

Unfortunately it's not really possible to cherrypick this back to 2.8, since there have been so many changes to the assignor since then. However the main conflicts are due to the fix #3, which isn't actually necessary as the bug it's fixing was only present in 3.0. So we just need to extract the #1 and #2 fixes and apply those to the 2.8 assignor (along with the tests, most if not all of which are still relevant in 2.8).

We should split this out into a separate ticket though, so we can close the current one and unblock the 3.0 release. @showuon If you have time would you be interested in picking this up? I guess an alternative is to just port those improvements you made to the assignor back to 2.8 and then just cherrypick this fix as usual, which should be relatively smooth since I doubt much else has touched the assignor recently. I'll leave it up to you to decide which route to go, if you want to take this one. Luckily there's no rush on this since the 2.8.1 release isn't right around the corner like the 3.0 release is, so we can take our time

**Comment by showuon** [comment]
Let me handle the porting to v2.8. :)

**Comment by ableegoldman** [comment]
@showuon  filed https://issues.apache.org/jira/browse/KAFKA-13081 just fyi

> Let me handle the porting to v2.8. :)

Awesome 🚀

**Pull request merged**
